pybricks/tools/pb_type_hostbuffer.c

```diff
@@ -0,0 +1,118 @@
+// SPDX-License-Identifier: MIT
+// Copyright (c) 2024 The Pybricks Authors
+
+#include "py/mpconfig.h"
+
+#if PYBRICKS_PY_TOOLS_HOSTBUFFER
+
+#include <string.h>
+
+#include <pbsys/command.h>
+
+#include "py/mphal.h"
+#include "py/objstr.h"
+
+#include <pybricks/tools.h>
+
+#include <pybricks/util_mp/pb_kwarg_helper.h>
+#include <pybricks/util_mp/pb_obj_helper.h>
+#include <pybricks/util_pb/pb_error.h>
+
+typedef struct _pb_type_hostbuffer_obj_t {
+    mp_obj_base_t base;
+    mp_obj_t format;
+    mp_obj_str_t bytes_obj;
+    uint8_t buffer[];
+} pb_type_hostbuffer_obj_t;
+
+// pointer to dynamically allocated hostbuffer singleton for driver callback.
+static pb_type_hostbuffer_obj_t *hostbuffer_instance;
+
+static void handle_write_data_buffer(uint16_t offset, uint32_t size, const uint8_t *data) {
+    // Can't write if buffer does not exist or isn't big enough.
+    if (!hostbuffer_instance || offset + size > hostbuffer_instance->bytes_obj.len) {
+        return;
+    }
+    memcpy(hostbuffer_instance->buffer + offset, data, size);
+}
+
+STATIC mp_obj_t pb_type_hostbuffer_get_bytes(mp_obj_t self_in) {
+    pb_type_hostbuffer_obj_t *self = MP_OBJ_TO_PTR(self_in);
+    // Don't return internal bytes object but make a copy so the user bytes
+    // object is constant as would be expected. Revisit: enable and return
+    // a memoryview, especially if using large buffers.
+    return mp_obj_new_bytes(self->bytes_obj.data, self->bytes_obj.len);
+}
+STATIC MP_DEFINE_CONST_FUN_OBJ_1(pb_type_hostbuffer_get_bytes_obj, pb_type_hostbuffer_get_bytes);
+
+STATIC mp_obj_t pb_type_hostbuffer_get_values(mp_obj_t self_in) {
+
+    // One-off import for ustruct.unpack.
+    static mp_obj_t ustruct_unpack = NULL;
```

Comment:
This seems a bit dangerous. It might work on bare-metal hubs if ustruct_unpack is an address in flash memory that is always the same. Otherwise, we could end up with a pointer to random data after restarting the MicroPython runtime.

If there isn't a measurable performance gain for caching this in a normal program, I would just always do the lookup and not worry about caching it.